api/envoy/api/v2/cds.proto

```diff
@@ -640,6 +640,23 @@ message Cluster {
   // outgoing connections made by the cluster. Order matters as the filters are
   // processed sequentially as connection events happen.
   repeated cluster.Filter filters = 40;
+
+  // If present, tells the client where to send load reports via LRS.
+  //
+  // If multiple clusters are configured to use the same LRS server,
+  // then the data for all of those clusters may be sent on the same LRS
+  // stream.
+  //
+  // Note that if LRS is configured this way, then there is no need for
+  // the LRS server to send LoadStatsResponse messages on the LRS
+  // stream, because the data it would have sent is already configured
+  // here.
+  message LoadReportConfig {
+    core.ApiConfigSource lrs_server = 1;
+    google.protobuf.Duration load_reporting_interval = 2;
+    bool report_endpoint_granularity = 3;
+  }
+  LoadReportConfig load_report_config = 41;
```

Comment:
Thanks, I'm largely supportive of adding this. There are a few things to sort out though:
1. We should have comments that are similar to https://github.com/envoyproxy/envoy/blob/817b2e36689cbc9e32892fc53f253f8e6361d5dd/api/envoy/config/bootstrap/v2/bootstrap.proto#L194.
2. Do we really want to duplicate bits of `LoadStatsResponse` in here? Should we just have an `initial_load_stats_response` message? Or should we just leave this to the bidi stream as exists today. Would be good to discuss this one further.
3. As per the other PR today, I'm wondering if someone is signed up to implement this in Envoy? We can't really take fields in the API like this that don't have an Envoy implementation.

Reply:
> 1. We should have comments that are similar to https://github.com/envoyproxy/envoy/blob/817b2e36689cbc9e32892fc53f253f8e6361d5dd/api/envoy/config/bootstrap/v2/bootstrap.proto#L194
>    .

Ah, interesting -- I didn't realize that Envoy had the LRS server configured via its bootstrap.  Should we deprecate that field in the bootstrap proto as part of this?  (See below for further discussion about migration strategy.)

I've added the requested comments.

> 2. Do we really want to duplicate bits of `LoadStatsResponse` in here? Should we just have an `initial_load_stats_response` message? Or should we just leave this to the bidi stream as exists today. Would be good to discuss this one further.

My thinking here was that this new mechanism would completely replace the existing mechanism, so when this new field is populated in CDS, there will never be a `LoadStatsResponse` message sent on the LRS stream.  Instead, for both initial responses and for subsequent changes, CDS will send an update of the cluster with the new values in these fields.  For example, to change the load reporting interval for a given cluster, CDS will send an update that includes the new load reporting interval.  Or, to stop sending load reports for a given cluster, CDS will send an update that leaves the `load_report_config` field unset.

However, thinking about this further, maybe we need to give more thought to the migration story between the current approach and the new approach.  I see two migration cases to consider here:

1. Envoy clients that do not yet support this new field.  These will ignore the new field even if the management server sends it, and they will require the management server to send the `LoadStatsResponse` message on the LRS stream.

2. Envoy clients that do support this new field but are talking to a management server that does not support the new field.  In this case, the client would need to continue configuring LRS via the bootstrap file as a fallback, and the management server would continue to send the `LoadStatsResponse` message on the LRS stream.

One possible way to handle this is to say that clients that start LRS based on the new CDS field will ignore any `LoadStatsResponse` sent on the LRS stream and document that management servers must continue sending `LoadStatsResponse` on the LRS stream until they are certain that there are no more old clients.

Another way to do this would be to add a new boolean field to the LRS request indicating whether the client was configured to use LRS via the new CDS field, in which case the LRS server will know not to send any responses on the LRS stream.  This is more explicit, but it does leave this new boolean field lying around even after the migration is complete, which seems sub-optimal.

Do we need to worry about federation cases where some clusters come from a management server that provides the new CDS field but others come from a management server that expects LRS to be configured via the bootstrap?  Or is that not a thing in the v2 APIs?

> 3. As per the other PR today, I'm wondering if someone is signed up to implement this in Envoy? We can't really take fields in the API like this that don't have an Envoy implementation.

I was not expecting to have to implement this in Envoy.  Envoy and gRPC are both consumers of this API, but if we agree jointly on an API change, I was assuming that it would be up to each of us to update our implementations to support the change.  If Envoy needs an API change in the future, I would not expect you folks to implement the change in gRPC.  And this approach certainly will not scale once there are other clients of this API.

We should probably discuss this further offline.  I do agree that this should be implemented in Envoy, and I don't want us to be in a situation where one client of this API is incentivized to say no to a change because they don't want to have to do any work to implement the change, nor do I want us to be in a situation where the ecosystem gets fragmented because new API additions get supported by only a subset of clients.  But I also don't want to be on the hook for updating Envoy whenever gRPC needs an API change.  So let's talk about this further to see if we can come up with a more sustainable approach.

Reply:
I think that we have two situations:
1. For v2 xDS, we should add the LRS field, to allow distinct config sources and servers per Cluster. This is directionally where we want to go. We can even specify an initial `LoadStatsResponse` here, that will be used by the client in the absence of any server message. We can deprecate in documentation etc. the streamed aspect here.
2. For v3 (which we can make happen at EOQ roughly), we can make the more radical changes.

How does that sound?

I agree we need to better coordinate on the implementation front. I don't think it should be expected that gRPC folks implement any API change on the Envoy side (although we wouldn't complain!), but simultaneously, we  need to schedule and budget engineering bandwidth in OSS or at Google to make the Envoy migration happen when we introduce deprecations like this; we can chat offline further about this.

Reply:
> To be clear, I don't actually have a use-case where I want to configure a different LRS server for different clusters.
> What I'm really after here is a way to get the LRS server via xDS, so that I don't have to configure it on the 
> client (e.g., via a bootstrap file). The CDS response seems like the appropriate place to put it in xDS, which means
> that it would inherently be able to be set on a per-cluster basis, even though that's not really what I'm after.

In an earlier comment you mentioned using a flag in LoadStatsRequest indicating that it was configured via CDS. This along with making the client ignore LoadStatsResponse from Management Server (if the Management Server has not yet implemented the flag) should cover the migration and compatibility story between the clients and server.
Including the example of two clusters that you have above.

Or, am I missing something,